### PR TITLE
pretty: use combining mark aware centering and wrapping

### LINKED ISSUE #9450: Not aligned Sympy's nice pritting of division
http://stackoverflow.com/q/30197808/1524913

I'm trying to nice print some divisions with Sympy but I noticed it didn't display aligned.

    import sympy
    sympy.init_printing(use_unicode=True)
    
    sympy.pprint(sympy.Mul(-1, sympy.Pow(-5, -1, evaluate=False), evaluate=False))
    # Output:
    # -1 
    # ───
    #  -5  # Note that "-5" is displayed slightly more on the right than "-1".

Reason/fix for this?

**EDIT:** I did a lot of reverse-engineering using `inspect.getsource` and `inspect.getsourcefile` but it didn't really help out in the end.

Pretty Printing in Sympy seems to be relying on the _Prettyprinter by Jurjen Bos_.

    import sympy
    
    from sympy.printing.pretty.stringpict import *
    
    sympy.init_printing(use_unicode=True)
    
    prettyForm("-1")/prettyForm("-5")
    # Displays:
    # -1
    # --
    # -5

So it does display aligned, but I can't get it to use unicode.

The PrettyPrinter is called from the file `sympy/printing/pretty/pretty.py` in the method `PrettyPrinter._print_Mul` which simply `return prettyForm.__mul__(*a)/prettyForm.__mul__(*b)` with, I thought, `a` and `b` simply being `['-1']` and `['-5']` but it wouldn't work.

### PR DESCRIPTION
**Pull request by pyBlob**
The console pprint-function does not insert the correct amount of spaces in an expression like:
```
>>> pprint([Symbol("ihat") / (Symbol("i") + 1)])
⎡  î ⎤
⎢─────⎥
⎣i + 1⎦
```

While it does for following expression:
```
>>> pprint(Matrix([Symbol("ihat"), Symbol("i") + 1]))
⎡  î  ⎤
⎢     ⎥
⎣i + 1⎦
```

The difference is that the centering done by `PrettyPrinter._print_matrix_contents` is unicode-aware, while the centering of `stringPict.equalLengths` and `stringPict.stack` is not.

The method `stringPict.render` also uses indexing to wrap lines. This messes up the alignments between individual lines of wrapped sub-pictures. This may even remove some of the trailing marks when the wrapping separates the character and its mark.

#### References to other Issues or PRs

- Fixes #9450

#### Brief description of what is fixed or changed

- Extract centering code of `PrettyPrinter._print_matrix_contents` to `pretty_symbology.center_pad`.
- Reuse for `stringPict.equalLengths` and `stringPict.stack` .
- Rewrite wrapping logic of `stringPict.render`:
  - Handle each character individually
  - Insert overflow indicators like "vim" and "less" do for long lines.

#### Other comments

TODO: Should I add the example at the top to the docstring of `center` or `center_pad`? Does it automatically run on test?

TODO: Fix broken whitespace in testcases + examples.

Why do the failing examples lack trailing whitespace? This PR should not have had any impact on those examples, because they don't use combining marks.

Example: (Select following code block and note the missing whitespace in "Expected")
```
sympy/sympy/solvers/ode/single.py", line 1634, in sympy.solvers.ode.single.HomogeneousCoeffSubsIndepDivDep
Failed example:
    pprint(dsolve(2*x*f(x) + (x**2 + f(x)**2)*f(x).diff(x), f(x),
    hint='1st_homogeneous_coeff_subs_indep_div_dep',
    simplify=False))
Expected:
                             /    2    \
                             | 3*x     |
                          log|----- + 1|
                             | 2       |
                             \f (x)    /
    log(f(x)) = log(C1) - --------------
                                3
Got:
                             /   2     \
                             |3*x      |
                          log|----- + 1|
                             | 2       |
                             \f (x)    /
    log(f(x)) = log(C1) - --------------
                                3       
```

#### Release Notes

* printing
  * pretty: use combining mark aware centering for equalLengths and stack
  * pretty: use combining mark aware wrapping

### CODE REVIEW

**Comment by oscargus** [comment]
Thank you for this contribution! Looks promising!

> Should I add the example at the top to the docstring of `center` or `center_pad`? 

I do not think it is needed.

> Does it automatically run on test?

Yes, it is executed as part of the doc-tests, but the main purpose of the doc-test is to make sure that the docs are correct, not that the function is correct. So please add tests in `printing/pretty/tests/test_pretty.py`.

File: sympy/printing/pretty/stringpict.py [hunk 1]
```diff
@@ -262,38 +259,63 @@ def render(self, * args, **kwargs):
...
-            i += ncols
-
-        if svals[-1] == '':
-            del svals[-1]  # Get rid of the last spacer
-
-        return "\n".join(svals)
+        """
+        Break long-lines in a visually pleasing format.
+        without overflow indicators | with overflow indicators
+
```
Comment (smichr) [inline sympy/printing/pretty/stringpict.py:272]:
```suggestion

```

File: sympy/printing/pretty/stringpict.py [hunk 2]
```diff
@@ -262,38 +259,62 @@ def render(self, * args, **kwargs):
...
-        return "\n".join(svals)
+        """
+        Break long-lines in a visually pleasing format.
+        without overflow indicators | with overflow indicators
+        |   2  2        3     |     |   2  2        3    ↪|
+        |6*x *y  + 4*x*y  +   |     |6*x *y  + 4*x*y  +  ↪|
+        |                     |     |                     |
+        |     3    4    4     |     |↪      3    4    4   |
+        |4*y*x  + x  + y      |     |↪ 4*y*x  + x  + y    |
+
```
Comment (smichr) [inline sympy/printing/pretty/stringpict.py:277]:
```suggestion

```

File: sympy/printing/pretty/tests/test_pretty.py [hunk 3]
```diff
@@ -3021,6 +3022,14 @@ def test_pretty_derivatives():
...
 dy dx             \
 """
+    ascii_str_3 = \
+"""\
+  2               \n\
+ d   / 2         \\\n\
+-----\\x  + log(x)/\n\
+dy dx             \
+"""
+
```
Comment (smichr) [inline sympy/printing/pretty/tests/test_pretty.py:3032]:
```suggestion
```

**Comment by smichr** [comment]
I will put up a replacement PR today that continues this.

**Pull request closed without merging**